haxe/ui/macros/ComponentMacros.hx

```diff
@@ -103,113 +102,112 @@ class ComponentMacros {
-        if (c.text != null)             s += 'c${id}.text = "${c.text}";\n';
-        if (c.styleNames != null)       s += 'c${id}.styleNames = "${c.styleNames}";\n';
-        if (c.style != null)            s += 'c${id}.styleString = "${c.styleString}";\n';
-        if (c.layoutName != null)       s += 'c${id}.layoutName = "${c.layoutName}";\n';
+        if (c.id != null)               assign("id", c.id);
+        if (c.left != null)             assign("left", c.left);
+        if (c.top != null)              assign("top", c.top);
+        if (c.width != null)            assign("width", c.width);
+        if (c.height != null)           assign("height", c.height);
+        if (c.percentWidth != null)     assign("percentWidth", c.percentWidth);
+        if (c.percentHeight != null)    assign("percentHeight", c.percentHeight);
+        if (c.text != null)             assign("text", c.text);
+        if (c.styleNames != null)       assign("styleNames", c.styleNames);
+        if (c.style != null)            assign("styleString", c.styleString);
+        if (c.layoutName != null)       assign("layoutName", c.layoutName);
         for (propName in c.properties.keys()) {
             var propValue = c.properties.get(propName);
-            if (propValue == "true" || propValue == "yes" || propValue == "false" || propValue == "no") {
-                propValue = '${propValue == "true" || propValue == "yes"}';
-            } else if (Std.parseInt(propValue) != null) {
-                propValue = '${Std.parseInt(propValue)}';
+            var propExpr = if (propValue == "true" || propValue == "yes" || propValue == "false" || propValue == "no") {
+                macro $v{propValue == "true" || propValue == "yes"};
             } else {
-                propValue = '"${propValue}"';
+                var intValue = Std.parseInt(propValue);
+                if (intValue != null) {
+                    macro $v{intValue};
+                } else {
+                    macro $v{propValue};
+                }
             }
 
             if (StringTools.startsWith(propName, "on")) {
-                s += 'c${id}.addScriptEvent("${propName}", ${propValue});\n';
+                add(macro $i{componentVarName}.addScriptEvent($v{propName}, $propExpr));
             } else {
-                s += 'c${id}.${propName} = ${propValue};\n';
+                add(macro $i{componentVarName}.$propName = $propExpr);
             }
         }
 
-
         if (MacroHelpers.hasInterface(type, "haxe.ui.core.IDataComponent") == true && c.data != null) {
-            s += 'cast(c${id}, haxe.ui.core.IDataComponent).dataSource = new haxe.ui.data.DataSourceFactory<Dynamic>().fromString("${c.dataString}", haxe.ui.data.ArrayDataSource);\n';
+            add(macro ($i{componentVarName} : haxe.ui.core.IDataComponent).dataSource = new haxe.ui.data.DataSourceFactory<Dynamic>().fromString($v{c.dataString}, haxe.ui.data.ArrayDataSource));
```

Comment:
Here I changed a `cast (c0, IDataComponent)` to `(c0 : IDataComponent)` because the former would have a run-time overhead, which generally only makes sense if the "Class cast exception" string is actually caught somewhere.